src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -888,70 +1019,221 @@ scaleCreditsForMerge SingleRun{} _ = ScaledCredits 0
 -- the sizes of the input runs. As as result, merge work would/could be more
 -- evenly distributed over time when the resident run is smaller than the worst
 -- case.
-scaleCreditsForMerge (MergingRun LevelLevelling _ _) c =
+scaleCreditsForMerge (MergingRun LevelLevelling _ _ _ _ _) (Credit c) =
     ScaledCredits (c * (1 + 4))
 -- A tiering merge has 5 runs at most (one could be held back to merged again)
 -- and must be completed before the level is full (once 4 more runs come in).
-scaleCreditsForMerge (MergingRun LevelTiering (NumRuns n) _) c =
+scaleCreditsForMerge (MergingRun LevelTiering (NumRuns n) _ _ _ _) (Credit c) =
     ScaledCredits ((c * n + 3) `div` 4)
     -- same as division rounding up: ceiling (c * n / 4)
 
-{-# SPECIALISE supplyMergeCredits :: ScaledCredits -> MergingRun IO h -> IO () #-}
--- TODO: implement doing merge werk in batches, instead of always taking the
--- MVar. The thresholds for doing merge work should be different for each level,
--- maybe co-prime?
+{-# SPECIALISE supplyMergeCredits :: ScaledCredits -> CreditThreshold -> MergingRun IO h -> IO () #-}
+-- | Supply the given amount of credits to a merging run. This /may/ cause an
+-- ongoing merge to progress.
+--
+-- See Note [Credits]
 supplyMergeCredits ::
-     (MonadSTM m, MonadST m, MonadMVar m, MonadMask m, MonadFix m)
+     forall m h. (MonadSTM m, MonadST m, MonadMVar m, MonadMask m, MonadFix m)
   => ScaledCredits
+  -> CreditThreshold
   -> MergingRun m h
   -> m ()
-supplyMergeCredits _ SingleRun{} = pure ()
-supplyMergeCredits (ScaledCredits c) (MergingRun _ _ var) = do
-    mergeIsDone <- withMVar var $ \case
+supplyMergeCredits _ _ SingleRun{} = pure ()
+supplyMergeCredits
+  (ScaledCredits c)
+  thresh@(CreditThreshold creditsThresh)
+  (MergingRun _ _ numInputEntries unspentCreditsVar totalStepsVar mergeVar) = do
+    -- Add credits to the total of unspent credits
+    Credit unspentCredits <- addUnspentCredits unspentCreditsVar (Credit c)
+    -- Compute our /estimate/ of what the new total of unspent credits is. The
+    -- /actual/ total might have been changed again by a different thread, but
+    -- that is okay because we will find out if\/when we run takeCredits.
+    let unspentCredits' = unspentCredits + c
+
+    stepsDone <- readPrimVar (getTotalStepsVar totalStepsVar)
+
+    -- The merge is already finished
+    if stepsDone >= unNumEntries numInputEntries then
+      pure ()
+    -- We can finish the merge immediately
+    else if stepsDone + unspentCredits' >= unNumEntries numInputEntries then do
+      isMergeDone <-
+        bracketOnError (takeAllUnspentCredits unspentCreditsVar)
+                       (putBackUnspentCredits unspentCreditsVar)
+                       (stepMerge mergeVar totalStepsVar)
+      when isMergeDone $ completeMerge mergeVar
```

Comment:
Also, would it be better switch the two branches, i.e. to first check whether we are over the threshold? If there are more credits than the threshold, we probably only want to do one uniform batch, even if we could complete the merge.

Reply:
Also, should we `assert isMergeDone` here? Ah no, it could be that some other thread already took the credits (leaving us with 0) and completed (or will complete) the merge. Interesting.

Reply:
> won't that mean there is a possibility that we don't finish the merge before the next expectCompletedMerge?

Ah, that's true. Then we probably need to merge the full remainder here even if it's more than a usual batch, which shouldn't happen often.